lib/spack/spack/platforms/cray.py

```diff
@@ -143,19 +144,24 @@ def _default_target_from_env(self):
                 env={'TERM': os.environ.get('TERM', '')},
                 output=str, error=os.devnull
             )
+
             default_from_module = ''.join(output.split())  # rm all whitespace
             if default_from_module:
                 tty.debug("Found default module:%s" % default_from_module)
                 return default_from_module
             else:
-                front_end = archspec.cpu.host().name
-                if front_end in list(
-                        map(lambda x: _target_name_from_craype_target_name(x),
-                            self._avail_targets())
-                ):
-                    tty.debug("default to front-end architecture")
-                    return archspec.cpu.host().name
+                front_end = archspec.cpu.host()
+                # Look for the frontend architecture or closest ancestor
+                # available in cray target modules
+                for front_end_possibility in [front_end] + front_end.ancestors:
+                    if front_end_possibility.name in list(
+                            map(lambda x: _target_name_from_craype_target_name(x),
+                                self._avail_targets())
+                    ):
+                        tty.debug("using front-end architecture or available ancestor")
+                        return front_end_possibility.name
```

Comment:
Non-quadratic version would be better:

```suggestion
                avail = [
                    _target_name_from_craype_target_name(x)
                    for x in self._avail_targets()
                ]
                for front_end_possibility in [front_end] + front_end.ancestors:
                    if front_end_possibility.name in avail:
                        tty.debug("using front-end architecture or available ancestor")
                        return front_end_possibility.name
```

Reply:
Good catch, thanks!